include/kframework/lemmas/bytes-simplification.k

```diff
@@ -68,71 +68,168 @@ module BYTES-SIMPLIFICATION [symbolic]
+
+    rule [range-inside-cHead-base-conc-right]:
+      #range(B:Bytes, S1:Int +Int S2:Int, W:Int) => #range(#range(B, S2, lengthBytes(B) -Int S2), S1, W)
+      requires 0 <=Int S1 andBool 0 <=Int S2 andBool 0 <Int S1 +Int S2 andBool S1 +Int S2 <=Int lengthBytes(B)
+      [concrete(B), concrete(S2), simplification]
+
+    rule [range-inside-cHead-concat]:
+      #range(B1:Bytes +Bytes B2:Bytes, S:Int, W:Int) => #range(#range(B1, S, lengthBytes(B1) -Int S) +Bytes B2, 0, W)
+      requires 0 <Int S andBool S <=Int lengthBytes(B1)
+      [simplification]
 
-    rule #range( M [ N := BUF:Bytes ], L, K ) =>
-         #let W = minInt(K, N -Int L) #in #range(M, L, W) +Bytes #range(M [ N := BUF ], N, K -Int W)
-      requires 0 <Int K andBool L <Int N [simplification]
+    rule [range-of-range]:
+      #range(#range(B:Bytes, S1:Int, W1:Int), S2:Int, W2:Int) =>
+        #range(B, S1 +Int S2, minInt(W2, W1 -Int S2)) +Bytes #buf(maxInt(0, W2 -Int (W1 -Int S2)), 0)
+      requires 0 <=Int S1 andBool 0 <=Int W1 andBool 0 <=Int S2 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range( M [ N := BUF:Bytes ], L, K ) =>
-           #let S = L -Int N #in
-           #let W = minInt(K, lengthBytes(BUF) -Int S) #in
-           #range(BUF, S, W) +Bytes #range(M, N +Int lengthBytes(BUF), K -Int W)
-      requires 0 <Int K andBool N <=Int L andBool L <=Int N +Int lengthBytes(BUF) [simplification]
+    rule [range-join-base]:
+      #range(B, S1, W1) +Bytes #range(B, S2, W2) => #range(B, S1, W1 +Int W2)
+      requires S2 ==Int S1 +Int W1
+       andBool 0 <=Int S1 andBool 0 <=Int W1 andBool 0 <=Int S2 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range( M [ N := BUF:Bytes ], L, K ) => #range(M, L, K)
-      requires 0 <Int K andBool N +Int lengthBytes(BUF) <=Int L [simplification]
+    rule [range-join-concat]:
+      #range(B1, S1, W1) +Bytes #range(B1, S2, W2) +Bytes B2 => #range(B1, S1, W1 +Int W2) +Bytes B2
+      requires S2 ==Int S1 +Int W1
+       andBool 0 <=Int S1 andBool 0 <=Int W1 andBool 0 <=Int S2 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range( M [ N1 := BUF1:Bytes ] [ N2 := BUF2:Bytes ], L, K ) => #range( M [ N2 := BUF2 ], L, K )
-        requires 0 <Int K andBool N1 +Int lengthBytes(BUF1) <=Int L [simplification]
+    rule [range-reform-base]:
+      #range(B:Bytes, S:Int, minInt(W1:Int, W2:Int)) +Bytes #buf(maxInt(0, W3:Int), 0) => #range (B, S, W1)
+      requires W2 ==Int lengthBytes(B) -Int S
+       andBool W3 ==Int S +Int W1 -Int lengthBytes(B)
+       andBool 0 <=Int S andBool 0 <=Int W1 andBool 0 <=Int W2
+      [simplification]
+
+    rule [range-reform-concat]:
+      #range(B1:Bytes, S:Int, minInt(W1:Int, W2:Int)) +Bytes #buf(maxInt(0, W3:Int), 0) +Bytes B2:Bytes => #range (B1, S, W1) +Bytes B2
+      requires W2 ==Int lengthBytes(B1) -Int S
+       andBool W3 ==Int S +Int W1 -Int lengthBytes(B1)
+       andBool 0 <=Int S andBool 0 <=Int W1 andBool 0 <=Int W2
+      [simplification]
```

Comment:
But this parenthetisation will be undone immediately by the right-associativity enforcement, will it not?

Reply:
Your lemma, as written, will be parsed as `(B1 +Bytes B2) +Bytes B3`. That doesn't have anything to do with the rules that enforce that our normal form is right-associative. Better to put the parens in so that people reading the lemma have no doubt about which AST you mean explicitly.

Reply:
I can't think of how to break them into simpler ones right now, will think a bit more.